merlin/systems/dag/dictarray.py

```diff
@@ -13,44 +13,151 @@
+            )
+
+    def cpu(self):
+        """
+        Move the data for this column to host (CPU) memory
+
+        Returns
+        -------
+        Column
+            Same column, same data but now definitely in CPU memory
+        """
+        self.device = Device.CPU
+        return self
+
+    def gpu(self):
+        """
+        Move the data for this column to device (GPU) memory
+
+        Returns
+        -------
+        Column
+            Same column, same data but now definitely in GPU memory
+        """
+        self.device = Device.GPU
+        return self
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, device):
+        if not cupy:
+            raise ValueError(
+                "Unable to move Column data between CPU and GPU without Cupy installed."
+            )
+
+        if device == "cpu":
+            device = Device.CPU
+        elif device == "gpu":
+            device = Device.GPU
+
+        # CPU to GPU
+        if self._device == Device.CPU and device == Device.GPU:
+            self._device_move(cupy.asarray)
+        # GPU to CPU
+        elif self._device == Device.GPU and device == Device.CPU:
+            self._device_move(cupy.asnumpy)
+        # Nothing to do
+        else:
+            return self
 
-        self.values = values
-        self.dtype = values.dtype
+    def _device_move(self, fn):
+        self.values = fn(self.values)
+        if self.row_lengths:
+            self.row_lengths = fn(self.row_lengths)
 
     def __getitem__(self, index):
-        return self.values[index]
+        if self.row_lengths:
+            start = self._array_lib.cumsum(self.row_lengths[:index])
+            end = start + self.row_lengths[index] - 1
+            return self.values[start:end]
+        else:
+            return self.values[index]
 
     def __eq__(self, other):
-        return all(self.values == other.values) and self.dtype == other.dtype
+        values_eq = all(self.values == other.values) and self.dtype == other.dtype
+        if self.row_lengths:
+            return values_eq and all(self.row_lengths == other.row_lengths)
+        else:
+            return values_eq
 
     def __len__(self):
-        return len(self.values)
+        if self.row_lengths:
+            return len(self.row_lengths)
+        else:
+            return len(self.values)
 
     @property
     def shape(self):
-        return self.values.shape
+        if self.row_lengths:
+            dim = self.row_lengths[0] if self.is_ragged else None
+            return (len(self), dim)
+        else:
+            return self.values.shape
+
+    @property
+    def is_list(self):
+        return len(self.values.shape) > 1 or self.row_lengths is not None
 
+    @property
+    def is_ragged(self):
+        return self.row_lengths and any(self.row_lengths != self.row_lengths[0])
```

Comment:
Based on our discussion yesterday. this property is consistent with the data in  `values/row_lengths`. However, for situations where the intermediate data is not ragged according to the row lengths, but does need to be ragged for the purposes of an operator. do we leave that up to an operator to do the coercion?